### PR TITLE
Apply ratchets to per kw charges

### PR DESCRIPTION
**Pull request by brtietz**
Pairs with SAM PR https://github.com/NREL/SAM/pull/728

- Renames billing demand variables to remove the reference to energy charges
- Adds variable "ur_dc_billing_demand_periods" to cover whether or or not a given demand charge TOU period should be included with billing demand
- Applies billing demand calculations to demand charges as well

Fixes https://github.com/NREL/SAM/issues/71

### CODE REVIEW

File: shared/lib_utility_rate_equations.cpp [hunk 1]
```diff
@@ -215,37 +219,64 @@ bool rate_data::check_for_kwh_per_kw_rate(int units) {
 
 double rate_data::get_billing_demand(int month) {
     int m = 0;
-    double billing_demand = ec_bd_minimum;
-    int prev_yr_lookback = 11 - (ec_bd_lookback_months - month); // What month do we stop looking back in the prev yr?
+    double billing_demand = bd_minimum;
+    int prev_yr_lookback = 11 - (bd_lookback_months - month); // What month do we stop looking back in the prev yr?
```
Comment (sjanzou) [inline shared/lib_utility_rate_equations.cpp:223]: How does this work across year boundaries - especially for year 1?

File: shared/lib_utility_rate_equations.cpp [hunk 2]
```diff
@@ -215,37 +219,64 @@ bool rate_data::check_for_kwh_per_kw_rate(int units) {
...
 
     return billing_demand;
 }
 
+void rate_data::set_billing_demands() {
+    for (int m = 0; m < (int) m_month.size(); m++) {
+        double flat_peak = m_month[m].dc_flat_peak;
+        if (en_billing_demand_lookback) {
+            // If ratchets are present the peak used here might be the actual peak, or something based on a previous month.
+            flat_peak = get_billing_demand(m);
```
Comment (sjanzou) [inline shared/lib_utility_rate_equations.cpp:274]: Is there a reference for this? i.e. are all flat demand charges to be based on the billing demand lookback? Seems like that overwrites the flat demand charge...

Comment (sjanzou) [reply shared/lib_utility_rate_equations.cpp:274]: Okay, you are not overwriting the ds_flat_peak - please disregard above comment.

File: ssc/common.cpp [hunk 3]
```diff
@@ -869,11 +869,12 @@ var_info vtab_utility_rate_common[] = {
...
+    { SSC_INPUT,        SSC_NUMBER,     "ur_billing_demand_lookback_period", "Billing demand lookback period",  "mn",         "",                "Electricity Rates",           "ur_enable_billing_demand=1",                 "INTEGER,MIN=0,MAX=12",                             "" },
+    { SSC_INPUT,        SSC_MATRIX,     "ur_billing_demand_lookback_percentages", "Billing demand lookback percentages by month and consider actual peak demand",       "",         "12x2",      "Electricity Rates",        "ur_enable_billing_demand=1",                 "",                             "" },
+    { SSC_INPUT,        SSC_MATRIX,     "ur_dc_billing_demand_periods", "Billing demand applicability to a given demand charge time of use period",       "",         "",      "Electricity Rates",        "ur_enable_billing_demand=1",                 "",                             "" },
+    { SSC_INPUT,        SSC_ARRAY,      "ur_yearzero_usage_peaks",  "Peak usage by month for year zero",       "",         "12",                "Electricity Rates",        "ur_enable_billing_demand=1",                 "",                             "" },
```
Comment (sjanzou) [inline ssc/common.cpp:877]: So, this takes care of the lookback across year 1 boundary. Should this be a user input or constucted from year 1 values? Maybe a questions for @cpaulgilman

File: test/ssc_test/cmod_utilityrate5_test.cpp [hunk 4]
```diff
@@ -1484,11 +1484,11 @@ TEST(cmod_utilityrate5_eqns, Test_Commercial_kWh_per_kW_charges_with_billing_dem
                                          11, 1, 9.9999999999999998e+37, 0 };
     ssc_data_set_matrix(data, "ur_dc_flat_mat", p_ur_dc_flat_mat, 12, 4);
 
-    ssc_data_set_number(data, "ur_ec_enable_billing_demand", 1);
-    ssc_data_set_number(data, "ur_ec_billing_demand_minimum", 100);
-    ssc_data_set_number(data, "ur_ec_billing_demand_lookback_period", 11);
+    ssc_data_set_number(data, "ur_enable_billing_demand", 1);
+    ssc_data_set_number(data, "ur_billing_demand_minimum", 100);
+    ssc_data_set_number(data, "ur_billing_demand_lookback_period", 11);
```
Comment (sjanzou) [inline test/ssc_test/cmod_utilityrate5_test.cpp:1489]: Do you have a min, max for lookback months? (Should be taken care of in the compute module precheck.

**Review by sjanzou — APPROVED** [review]
Some comments, but looks good!

File: test/ssc_test/cmod_utilityrate5_test.cpp (hunk 4 above)
Comment (brtietz) [reply test/ssc_test/cmod_utilityrate5_test.cpp:1489]: Yes, 0 and 12:

image

**Pull request merged**